rust/noosphere-gateway/src/jobs/worker_queue/queue.rs

```diff
@@ -0,0 +1,70 @@
+use super::{queue_thread::WorkerQueueThread, Processor};
+use anyhow::Result;
+use std::{sync::Arc, time::Duration};
+use tokio::{
+    sync::mpsc::{unbounded_channel, UnboundedSender},
+    task::JoinHandle,
+};
+
+/// An abstraction around managing several worker threads, and distributing
+/// work amongst them.
+///
+/// To terminate all processing, all references to the [WorkerQueue]
+/// must be dropped.
+#[derive(Debug, Clone)]
+pub struct WorkerQueue<P: Processor> {
+    handle: Option<Arc<JoinHandle<Result<()>>>>,
+    request_tx: UnboundedSender<P::Job>,
+}
+
+impl<P> WorkerQueue<P>
+where
+    P: Processor,
+{
+    /// Creates a new [WorkerQueue] and starts its worker threads.
+    ///
+    /// By default, `retries` is set to 1 and `timeout` is 3 minutes.
+    pub fn spawn(
+        worker_count: usize,
+        worker_context: P::Context,
+        retries: Option<usize>,
+        timeout: Option<Duration>,
+    ) -> Self {
+        let (request_tx, request_rx) = unbounded_channel();
+        let handle = Some(Arc::new(tokio::spawn(async move {
+            let workers = WorkerQueueThread::<P>::new(
```

Comment:
The naming and documentation were confusing to read the first time around. `WorkerQueue` sounds like "the thing that has a pool of workers." In fact, the documentation currently reads:

```rs
/// An abstraction around managing several worker threads, and distributing
/// work amongst them.
```

And, there is `WorkerQueueThread`, which sounds like the thing that does work on a thread. But, most of the meaningful management of multiple threads is performed by `WorkerQueueThread`.

Why are these separate constructs? It kind of looks like all that `WorkerQueue` does is thinly wrap `WorkerQueueThread`. Also: can we call `WorkerQueueThread` something closer to what it is? Maybe it's just `WorkerQueue`, maybe it's a `WorkerQueuePool`. It's certainly not a single thread so long as it has a `Vec<Workers<P>>`.